token/program-2022/src/extension/confidential_transfer/processor.rs

```diff
@@ -450,139 +452,329 @@ fn process_transfer(
     let authority_info = next_account_info(account_info_iter)?;
-    let authority_info_data_len = authority_info.data_len();
 
     check_program_account(mint_info.owner)?;
     let mint_data = &mint_info.data.borrow_mut();
     let mint = StateWithExtensions::<Mint>::unpack(mint_data)?;
     let ct_mint = mint.get_extension::<ConfidentialTransferMint>()?;
 
+    let previous_instruction =
+        get_instruction_relative(proof_instruction_offset, instructions_sysvar_info)?;
+
+    if let Ok(transfer_fee_config) = mint.get_extension::<TransferFeeConfig>() {
+        // mint is extended for fees
+        let proof_data = decode_proof_instruction::<TransferWithFeeData>(
+            ProofInstruction::VerifyTransfer,
+            &previous_instruction,
+        )?;
+
+        if proof_data.transfer_with_fee_pubkeys.auditor != ct_mint.auditor_pk {
+            return Err(TokenError::ConfidentialTransferElGamalPubkeyMismatch.into());
+        }
+
+        // `withdraw_withheld_authority` ElGamal pubkey in proof data and mint must match
+        if proof_data.transfer_with_fee_pubkeys.fee_collector
+            != ct_mint.withdraw_withheld_authority_pk
+        {
+            return Err(TokenError::ConfidentialTransferElGamalPubkeyMismatch.into());
+        }
+
+        // fee parameters in proof data and mint must match
+        let epoch = Clock::get()?.epoch;
+        let (maximum_fee, transfer_fee_basis_points) =
+            if u64::from(transfer_fee_config.newer_transfer_fee.epoch) < epoch {
+                (
+                    u64::from(transfer_fee_config.older_transfer_fee.maximum_fee),
+                    u16::from(
+                        transfer_fee_config
+                            .older_transfer_fee
+                            .transfer_fee_basis_points,
+                    ),
+                )
+            } else {
+                (
+                    u64::from(transfer_fee_config.newer_transfer_fee.maximum_fee),
+                    u16::from(
+                        transfer_fee_config
+                            .newer_transfer_fee
+                            .transfer_fee_basis_points,
+                    ),
+                )
+            };
+
+        if u64::from(proof_data.fee_parameters.maximum_fee) != maximum_fee
+            || u16::from(proof_data.fee_parameters.fee_rate_basis_points)
+                != transfer_fee_basis_points
+        {
+            return Err(TokenError::FeeParametersMismatch.into());
+        }
+
+        // Process source account
+        process_source_for_transfer(
+            program_id,
+            token_account_info,
+            mint_info,
+            authority_info,
+            account_info_iter.as_slice(),
+            &proof_data.transfer_with_fee_pubkeys.source,
+            (
+                proof_data.ciphertext_lo.commitment,
+                proof_data.ciphertext_lo.source,
+            )
+                .into(),
```

Comment:
nit: this formatting is a little weird.  Did `cargo fmt` do this?

Reply:
Yes, it is quite weird. If I untab `into()`, `cargo fmt` seems to complain. Even if I make it a separate variable, it seems to complain. I changed it to use `from` instead as in the original code. I think it should be more readable now.